// @zenvia/sdk/dist/lib/templates/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Template = void 0;
/**
 * Implementation of base templates.
 */
class Template {
    /**
     * Returns a new `Template` that is used to create a new template.
     *
     * @param name The name to identifier the template.
     * @param locale The language and locale.
     * @param channel The channel who template will be used.
     * @param category The category used to identifier in WhatsApp.
     * @param senderId This is the identifier of sender for this template. The sender shoud be created with a credential.
     * @param notificationEmail Mail list (comma-separated) to send notifications about the message template approving process.
     * @param components An [[IComponents]] object.
     */
    constructor(name, locale, channel, category, senderId, notificationEmail, components) {
        this.name = name;
        this.locale = locale;
        this.channel = channel;
        this.category = category;
        this.senderId = senderId;
        this.notificationEmail = notificationEmail;
        this.components = components;
    }
}
exports.Template = Template;
