// @zenvia/sdk/dist/lib/contents/location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocationContent = void 0;
const abstract_content_1 = require("./abstract-content");
/**
 * Implementation of location content.
 */
class LocationContent extends abstract_content_1.AbstractContent {
    /**
     * Returns a new `LocationContent` that can be used to send location to your customer.
     *
     * @param longitude Longitude of the location.
     * @param latitude Latitude of the location.
     * @param name Name of the location.
     * @param address Address of the location. Only displayed if name is present.
     * @param url URL for the website where the location information was downloaded. (WhatsApp restriction: available only on IN direction)
     */
    constructor(longitude, latitude, name, address, url) {
        super('location');
        this.longitude = longitude;
        this.latitude = latitude;
        this.name = name;
        this.address = address;
        this.url = url;
    }
}
exports.LocationContent = LocationContent;
